### ISSUE TITLE
T-SNE visualization of the learned feature embeddings looks different

### ISSUE THREAD

**Issue by jpainam** [vana77/Bottom-up-Clustering-Person-Re-identification#24]
Hi; I tried to  reproduce and visualize the 2-dimension feature map of the clustering using tSNE . This is the output. 
![Figure_1](https://user-images.githubusercontent.com/1701705/82729212-37e18280-9d28-11ea-907d-042c9373f9c6.png)

It looks very different to your figure 4 even in the 18th step .
This is what I did to get the feature map at each step.
In `get_new_train_data` I saved `u_feat` and the labels. The `u_feat` is a `12936x2048` where each line represent a image and it's extracted features. I'm able to randomly choose 100 identities based on `new_train_data` ([tuple(images), pid, camid, video_id])

```python
def get_new_train_data(self, labels, nums_to_merge, size_penalty, step):
   u_feas, feature_avg, label_to_images, fc_avg = self.generate_average_feature(labels)
   ..... 
   # before return, save
   np.savez("logs/clusters_{}.npz".format(step), u_feas, labels, new_train_data)
```

**Comment by vana77** [comment]
Hi, the label used in tSNE should be the ground truth label, so that you can find if images of the same person are gathered together. However, "labels" you used in the code is the estimated label by clustering. 
The visualization I drew was like below:
![visual](https://user-images.githubusercontent.com/26541240/82734972-e5688c00-9d50-11ea-839a-4576bc35df32.png)

**Comment by jpainam** [comment]
Thank you. you are right, the results are much more better now.

**Comment by jpainam** [comment]
For confirmation sake
```python
tsne = TSNE(n_components=2, verbose=1, n_iter=1000)
    embedded = tsne.fit_transform(features, y=new_labels)
    data = {'x': embedded[:, 0],
            'labels': new_labels,
            'y': embedded[:, 1]}

    ax.set_xlabel(xlabel)
    sns.scatterplot(x="x", y="y",
                    hue="labels",
                    palette=sns.color_palette("hls", len(np.unique(new_labels))),
                    data=data,
                    legend=False,
                    alpha=1.0,
                    linewidth=0,
                    s=10,
                    ax=ax)
```
![image](https://user-images.githubusercontent.com/1701705/82830204-aae52780-9ee7-11ea-8479-49edf36b863f.png)

**Comment by MateRyze** [comment]
> For confirmation sake
> 
> ```python
> tsne = TSNE(n_components=2, verbose=1, n_iter=1000)
>     embedded = tsne.fit_transform(features, y=new_labels)
>     data = {'x': embedded[:, 0],
>             'labels': new_labels,
>             'y': embedded[:, 1]}
> 
>     ax.set_xlabel(xlabel)
>     sns.scatterplot(x="x", y="y",
>                     hue="labels",
>                     palette=sns.color_palette("hls", len(np.unique(new_labels))),
>                     data=data,
>                     legend=False,
>                     alpha=1.0,
>                     linewidth=0,
>                     s=10,
>                     ax=ax)
> ```
> 
> ![image](https://user-images.githubusercontent.com/1701705/82830204-aae52780-9ee7-11ea-8479-49edf36b863f.png)

Which labels (ground truth data) and features have you used? Do they come from the function input `get_new_train_data`?

**Comment by vana77** [comment]
Hi @MateRyze, 
labels: self.u_label 
features: u_feas

**Comment by MateRyze** [comment]
Thank you @vana77 :)

**Comment by xkxkx** [comment]
Hello, is it convenient to provide t-sne visualization code?
